### DISCUSSION TITLE
How to use pre-trained model's learnable encoder/decoder for Conv-TasNet

### DISCUSSION THREAD

**Discussion by cemox35** [asteroid-team/asteroid#583] [Q&A] [answered]
Hello,

I trained the Conv-TasNet using default parameters.  I want to use this pre-trained model's encoder and decoder in different Conv-TasNet model(I just change the separation module). Can I do that?

**Comment by mpariente** [comment]
Yes, do you want them to be fixed or keep learning?

**Reply by cemox35** [reply]
Thank you for your answer. I want to fix them. How can I do that?

**Comment by JorisCos** [comment, marked as answer]
I am going to write a small example to show how to proceed, as it is a nice thing to have in your toolbox.

**Reply by JorisCos** [reply]
```python
import torch
from pytorch_lightning import Trainer

from asteroid.data import LibriMix
from asteroid.engine import System
from asteroid.losses import PITLossWrapper, pairwise_neg_sisdr
from asteroid.models import BaseModel, DPRNNTasNet

# Import your model (could be a pretrained from hugging face or a custom one)
model_pretrained = BaseModel.from_pretrained("JorisCos/ConvTasNet_Libri2Mix_sepnoisy_8k")

# Instantiate a new model, here we change from a TCN to a DPRNN module for separation
model = DPRNNTasNet(n_src=2, n_filters=512, n_repeats=1)


# We have to transfer the weights of the encoder/decoder to the new model
model.encoder = model_pretrained.encoder
model.decoder = model_pretrained.decoder

# Now freeze the encoder/decoder
for param in model.encoder.parameters():
    param.requires_grad = False
for param in model.decoder.parameters():
    param.requires_grad = False

# Train as usual
train_loader, val_loader = LibriMix.loaders_from_mini(task="sep_clean", batch_size=16)
optimizer = torch.optim.Adam(model.parameters())
loss_func = PITLossWrapper(pairwise_neg_sisdr, pit_from="pw_mtx")
system = System(model, optimizer, loss_func, train_loader, val_loader)
trainer = Trainer(max_epochs=1, gpus=0, limit_train_batches=0.03)
trainer.fit(system)
```

**Reply by JorisCos** [reply]
Just note that you have to make sure that your masker expects data with the right shape, i.e the output shape of your encoder. 
In my example, that's why I instantiate  `DPPRNTasNet` with `n_filters=512`

**Reply by JorisCos** [reply]
if you run this example you can look at the state of the encoder/decoder with `model.encoder.state_dict()['filterbank._filters']` and see that nothing changes after training while the weights in `model.masker.state_dict()` have changed.

**Reply by cemox35** [reply]
Thank you for your answer.  I want to load my model(not from_pretrained(), actually), but I cannot load my auto-checkpoint(in lightning-logs). Could you help that problem? I am very grateful.

**Reply by JorisCos** [reply]
If you trained with Asteroid you should be able to run `ConvTasNet.from_pretrained("best_model.pth")`

**Reply by JorisCos** [reply]
If you just have a `.ckpt` you could try something like this : 
```python
state_dict = torch.load(os.path.join(exp_dir,'your_path_to_your_log.ckpt'))
system.load_state_dict(state_dict=state_dict["state_dict"])
system.cpu()

to_save = system.model.serialize()
torch.save(to_save, os.path.join(exp_dir, "model.pth"))
```

**Reply by cemox35** [reply]
Thank you! It works.
